paasta_tools/cli/cmds/mark_for_deployment.py

```diff
@@ -1107,293 +1112,187 @@ def log_user_rollback(self) -> None:
         )
 
 
-class ClusterData:
-    """An auxiliary data transfer class.
-
-    Used by _query_clusters(), instances_deployed(),
-    _run_cluster_worker(), _run_instance_worker().
-
-    :param cluster: the name of the cluster.
-    :param service: the name of the service.
-    :param git_sha: git sha marked for deployment.
-    :param instances_queue: a thread-safe queue. Should contain all cluster
-                            instances that need to be checked.
-    :type instances_queue: Queue
-    """
-
-    def __init__(self, cluster, service, git_sha, instances_queue):
-        self.cluster = cluster
-        self.service = service
-        self.git_sha = git_sha
-        self.instances_queue = instances_queue
+async def wait_until_instance_is_done(
+    pool,
+    service: str,
+    instance: str,
+    cluster: str,
+    git_sha: str,
+    instance_config: LongRunningServiceConfig,
+) -> Tuple[str, str]:
+    loop = asyncio.get_running_loop()
+    while not await loop.run_in_executor(
+        pool,
+        functools.partial(
+            check_if_instance_is_done,
+            service,
+            instance,
+            cluster,
+            git_sha,
+            instance_config,
+        ),
+    ):
+        await asyncio.sleep(5)
+    return (
+        cluster,
+        instance,
+    )  # for the convenience of the caller, to know which future is finishing.
+
+
+def check_if_instance_is_done(
+    service: str,
+    instance: str,
+    cluster: str,
+    git_sha: str,
+    instance_config: LongRunningServiceConfig,
+    api: Optional[client.PaastaOApiClient] = None,
+) -> bool:
+    if api is None:
+        api = client.get_paasta_oapi_client(cluster=cluster)
+        if not api:
+            log.warning(
+                "Couldn't reach the PaaSTA api for {}! Assuming it is not "
+                "deployed there yet.".format(cluster)
+            )
+            return False
 
-    def __repr__(self):
-        return (
-            f"ClusterData(cluster={self.cluster}, service={self.service}, "
-            f"git_sha={self.git_sha}, instances_queue={self.instances_queue})"
+    log.debug(
+        "Inspecting the deployment status of {}.{} on {}".format(
+            service, instance, cluster
         )
-
-
-def instances_deployed(cluster_data, instances_out, green_light):
-    """Create a thread pool to run _run_instance_worker()
-
-    :param cluster_data: an instance of ClusterData.
-    :param instances_out: a empty thread-safe queue. I will contain
-                          instances that are not deployed yet.
-    :type instances_out: Queue
-    :param green_light: See the docstring for _query_clusters().
-    """
-    num_threads = min(5, cluster_data.instances_queue.qsize())
-
-    workers_launched = []
-    for _ in range(num_threads):
-        worker = Thread(
-            target=_run_instance_worker, args=(cluster_data, instances_out, green_light)
+    )
+    try:
+        status = None
```

Comment:
This is how it looked before. I think the intent here was something like

```
try:
    status = api.service.status_instance(...)
except api.api_error as e:
    ...
    status = None
```

because there's some code later that checks if status is None.